mlkem/src/native/riscv64/src/rv64v_poly.c

```diff
@@ -0,0 +1,750 @@
+{
+  size_t vl = MLK_RVV_E16M1_VL;
+  size_t vl2 = 2 * vl;
+
+  const vuint16m2_t v2p8 = bitswap_perm(3, 4, vl2);
+  const vuint16m2_t v2p4 = bitswap_perm(2, 4, vl2);
+  const vuint16m2_t v2p2 = bitswap_perm(1, 4, vl2);
+
+  /*    p0 = p2(p4(p8)) */
+  const vuint16m2_t v2p0 = __riscv_vrgather_vv_u16m2(
+      __riscv_vrgather_vv_u16m2(v2p8, v2p4, vl2), v2p2, vl2);
+
+  const vuint16m1_t vid = __riscv_vid_v_u16m1(vl);
+  const vuint16m1_t cs8 =
+      __riscv_vadd_vx_u16m1(__riscv_vsrl_vx_u16m1(vid, 3, vl), 2, vl);
+  const vuint16m1_t cs4 =
+      __riscv_vadd_vx_u16m1(__riscv_vsrl_vx_u16m1(vid, 2, vl), 2 + 2, vl);
+  const vuint16m1_t cs2 =
+      __riscv_vadd_vx_u16m1(__riscv_vsrl_vx_u16m1(vid, 1, vl), 2 + 2 + 4, vl);
+
+  vint16m1_t t0, t1, c0, vt;
+
+  /*    initial permute */
+  vp = __riscv_vrgatherei16_vv_i16m2(vp, v2p0, vl2);
+  t0 = __riscv_vget_v_i16m2_i16m1(vp, 0);
+  t1 = __riscv_vget_v_i16m2_i16m1(vp, 1);
+  c0 = __riscv_vrgather_vv_i16m1(cz, cs2, vl);
+  MLK_RVV_BFLY_RV(t0, t1, vt, c0, vl);
+
+  /*    swap 2  */
+  vp = __riscv_vcreate_v_i16m1_i16m2(t0, t1);
+  vp = __riscv_vrgatherei16_vv_i16m2(vp, v2p2, vl2);
+  t0 = __riscv_vget_v_i16m2_i16m1(vp, 0);
+  t1 = __riscv_vget_v_i16m2_i16m1(vp, 1);
+  c0 = __riscv_vrgather_vv_i16m1(cz, cs4, vl);
+  MLK_RVV_BFLY_RV(t0, t1, vt, c0, vl);
+
+  /*    swap 4  */
+  vp = __riscv_vcreate_v_i16m1_i16m2(t0, t1);
+  vp = __riscv_vrgatherei16_vv_i16m2(vp, v2p4, vl2);
+  t0 = __riscv_vget_v_i16m2_i16m1(vp, 0);
+  t1 = __riscv_vget_v_i16m2_i16m1(vp, 1);
+  c0 = __riscv_vrgather_vv_i16m1(cz, cs8, vl);
+  MLK_RVV_BFLY_RV(t0, t1, vt, c0, vl);
+
+  /*    swap 8  */
+  vp = __riscv_vcreate_v_i16m1_i16m2(t0, t1);
+  vp = __riscv_vrgatherei16_vv_i16m2(vp, v2p8, vl2);
+  t0 = __riscv_vget_v_i16m2_i16m1(vp, 0);
+  t1 = __riscv_vget_v_i16m2_i16m1(vp, 1);
+
+  /*    normalize   */
+  t0 = fq_mulq_vx(t0, MLK_RVV_MONT_R1, vl);
+  t1 = fq_mulq_vx(t1, MLK_RVV_MONT_R1, vl);
+
+  vp = __riscv_vcreate_v_i16m1_i16m2(t0, t1);
+
+  return vp;
+}
+
+void mlk_rv64v_poly_invntt_tomont(int16_t *r)
+{
+/* zetas can be compiled into vector constants; don't pass as a pointer */
+#include "rv64v_izetas.inc"
+
+  size_t vl = MLK_RVV_E16M1_VL;
+  size_t vl2 = 2 * vl;
+
+  const vint16m1_t z0 = __riscv_vle16_v_i16m1(&izeta[0x00], vl);
+  const vint16m1_t z2 = __riscv_vle16_v_i16m1(&izeta[0x10], vl);
+  const vint16m1_t z4 = __riscv_vle16_v_i16m1(&izeta[0x20], vl);
+  const vint16m1_t z6 = __riscv_vle16_v_i16m1(&izeta[0x30], vl);
+  const vint16m1_t z8 = __riscv_vle16_v_i16m1(&izeta[0x40], vl);
+  const vint16m1_t za = __riscv_vle16_v_i16m1(&izeta[0x50], vl);
+  const vint16m1_t zc = __riscv_vle16_v_i16m1(&izeta[0x60], vl);
+  const vint16m1_t ze = __riscv_vle16_v_i16m1(&izeta[0x70], vl);
+
+  vint16m1_t vt;
+  vint16m1_t v0, v1, v2, v3, v4, v5, v6, v7, v8, v9, va, vb, vc, vd, ve, vf;
+  vint16m2_t vp;
+
+  vp = mlk_rv64v_intt2(__riscv_vle16_v_i16m2(&r[0x00], vl2), z0);
```

Comment:
Blocker: We currently make no bounds assumption on the inputs, which forces an implementation to either reduce the inputs upfront, or frontload the Montgomery normalization (which sufficiently reduces as a side-effect).